x/operator/keeper/slash.go

```diff
@@ -15,224 +17,283 @@ import (
-				TotalAmount: slashInfo.AmountFromOptedIn.Neg(),
-			})
+			err = k.delegationKeeper.SetStakerShareToZero(ctx, parameter.Operator.String(), assetID, stakerMap)
 			if err != nil {
 				return err
 			}
-			// decrease the related share value
-			err = k.UpdateOptedInAssetsState(ctx, stakerID, assetID, operatorAddress.String(), slashInfo.AmountFromOptedIn.Neg())
+			err = k.delegationKeeper.DeleteStakerMapForOperator(ctx, parameter.Operator.String(), assetID)
 			if err != nil {
 				return err
 			}
+			changeAmount.TotalShare = sdkmath.LegacyNewDec(0)
+			changeAmount.OperatorAmount = sdkmath.NewInt(0)
+			changeAmount.OperatorShare = sdkmath.LegacyNewDec(0)
 		}
-	}
-	return nil
-}
 
-func (k *Keeper) SlashOperator(ctx sdk.Context, operatorAddress sdk.AccAddress, slashOperatorInfo map[string]*slashAmounts, processedHeight uint64) error {
-	for assetID, slashInfo := range slashOperatorInfo {
-		slashSumValue := slashInfo.AmountFromUnbonding.Add(slashInfo.AmountFromOptedIn)
-		// handle the state that needs to be updated when slashing both opted-in and unbonding assets
-		err := k.assetsKeeper.UpdateOperatorAssetState(ctx, operatorAddress, assetID, assetstype.OperatorSingleAssetChangeInfo{
-			TotalAmount:                        slashSumValue.Neg(),
-			OperatorAmount:                     slashInfo.AmountFromOptedIn.Neg(),
-			OperatorUnbondableAmountAfterSlash: slashInfo.AmountFromUnbonding.Neg(),
-		})
-		if err != nil {
-			return err
-		}
-		// Record the slash information for scheduled tasks and send it to the client chain once the veto duration expires.
-		err = k.UpdateSlashAssetsState(ctx, assetID, operatorAddress.String(), processedHeight, slashSumValue)
-		if err != nil {
-			return err
-		}
-
-		// handle the state that needs to be updated when slashing opted-in assets
-		// decrease the related share value
-		err = k.UpdateOptedInAssetsState(ctx, "", assetID, operatorAddress.String(), slashInfo.AmountFromOptedIn.Neg())
+		err = k.assetsKeeper.UpdateOperatorAssetState(ctx, parameter.Operator, assetID, changeAmount)
 		if err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
-// Slash The occurredSateHeight should be the height that has the latest stable state.
-func (k *Keeper) Slash(ctx sdk.Context, operatorAddress sdk.AccAddress, avsAddr, slashContract, slashID string, slashEventHeight int64, slashProportion sdkmath.LegacyDec) error {
-	height := ctx.BlockHeight()
-	if slashEventHeight > height {
-		return errorsmod.Wrap(types.ErrSlashOccurredHeight, fmt.Sprintf("slashEventHeight:%d,curHeight:%d", slashEventHeight, height))
-	}
-
-	// get the state when the slash occurred
-	// get the opted-in info
-	historicalSateCtx, err := assetstype.ContextForHistoricalState(ctx, slashEventHeight)
+// Slash performs all slash events include instant slash and no-instantaneous slash.
+func (k *Keeper) Slash(ctx sdk.Context, parameter *SlashInputInfo) error {
```

Comment:
The handling of different types of slash mechanisms (instant and non-instant slash) can be refactored using the **Strategy Pattern**. The Strategy Pattern allows the implementation of an algorithm or operation to be chosen at runtime so that the algorithm changes independently of the client using it. 

1. First, define an interface (Strategy Interface) that represents the different punishment strategies, 
2. and then implement this interface for each punishment mechanism. 
3. Finally, create a Context class that will select and use the appropriate strategy for each situation.

Reply:
In the new version of golang, it can use generic feature, also can apply in updateAssetValue with different type:
`func UpdateValue[T any](valueToUpdate, changeValue *T) error{} `

Reply:
Seems like the generic type can't be used to reduce the repeated code because `math.LegacyDec` and `math.Int` have some customized functions, such as `Neg() (res Int)`, which can't be handled by generic types or interfaces.

Reply:
> The handling of different types of slash mechanisms (instant and non-instant slash) can be refactored using the **Strategy Pattern**. The Strategy Pattern allows the implementation of an algorithm or operation to be chosen at runtime so that the algorithm changes independently of the client using it.
> 
> 1. First, define an interface (Strategy Interface) that represents the different punishment strategies,
> 2. and then implement this interface for each punishment mechanism.
> 3. Finally, create a Context class that will select and use the appropriate strategy for each situation.

Only one slash might remain, depending on the new slash design implementation. So the change should be considered in light of the new slash implementation.